### ISSUE TITLE
Language Server Crash in Component With Classes

### ISSUE THREAD

**Issue by Zehua-Chen** [vuejs/vetur#2254]
- [x] I have searched through existing issues
- [x] I have read through [docs](https://vuejs.github.io/vetur)
- [x] I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)

## Info

- Platform: macOS
- Vetur version: 0.27.1
- VS Code version: 1.48.2

## Problem

After opening vue files containing components with classes, the language server crashes

```
Loaded bundled typescript@4.0.2.
Vetur initialized
/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:364
    if (property.valueDeclaration.kind !== checkSyntaxKind) {
                                  ^

TypeError: Cannot read property 'kind' of undefined
    at getPropertyDecoratorNames (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:364:35)
    at /Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:87:33
    at Array.filter (<anonymous>)
    at getClassProps (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:87:14)
    at getClassAndObjectInfo (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:351:35)
    at getProps (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:65:20)
    at analyzeDefaultExportExpr (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:39:19)
    at Object.getComponentInfo (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/componentInfo.js:19:25)
    at Object.updateFileInfo (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/modes/script/javascript.js:65:46)
    at /Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/server/dist/services/vueInfoService.js:18:24
[Error - 11:33:51 PM] Connection to server got closed. Server will not be restarted.
[Error - 11:33:51 PM] Request textDocument/codeAction failed.
Error: Connection got disposed.
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:904:25)
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:74:35)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2309:42)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/main.js:155:15)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2296:18)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:240:26)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at IPCMessageReader.fireClose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at ChildProcess.<anonymous> (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:213:45)
	at ChildProcess.emit (events.js:208:15)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
[Error - 11:33:51 PM] Request textDocument/documentLink failed.
Error: Connection got disposed.
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:904:25)
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:74:35)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2309:42)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/main.js:155:15)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2296:18)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:240:26)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at IPCMessageReader.fireClose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at ChildProcess.<anonymous> (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:213:45)
	at ChildProcess.emit (events.js:208:15)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
[Error - 11:33:51 PM] Request textDocument/colorPresentation failed.
Error: Connection got disposed.
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:904:25)
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:74:35)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2309:42)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/main.js:155:15)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2296:18)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:240:26)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at IPCMessageReader.fireClose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at ChildProcess.<anonymous> (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:213:45)
	at ChildProcess.emit (events.js:208:15)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
[Error - 11:33:51 PM] Request textDocument/foldingRange failed.
Error: Connection got disposed.
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:904:25)
	at Object.dispose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:74:35)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2309:42)
	at LanguageClient.handleConnectionClosed (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/main.js:155:15)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-languageclient/lib/client.js:2296:18)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at closeHandler (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/main.js:240:26)
	at CallbackList.invoke (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at IPCMessageReader.fireClose (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at ChildProcess.<anonymous> (/Users/ZehuaChen/.vscode/extensions/octref.vetur-0.27.1/node_modules/vscode-jsonrpc/lib/messageReader.js:213:45)
	at ChildProcess.emit (events.js:208:15)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
```

## Reproducible Case

1. Create .tsconfig

```json
{
  "compilerOptions": {
    /* Visit https://aka.ms/tsconfig.json to read more about this file */

    /* Basic Options */
    // "incremental": true,                   /* Enable incremental compilation */
    "target": "ESNext",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
    "module": "ESNext",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
    // "lib": [],                             /* Specify library files to be included in the compilation. */
    // "allowJs": true,                       /* Allow javascript files to be compiled. */
    // "checkJs": true,                       /* Report errors in .js files. */
    // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
    // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
    // "sourceMap": true,                     /* Generates corresponding '.map' file. */
    // "outFile": "./",                       /* Concatenate and emit output to single file. */
    // "outDir": "./",                        /* Redirect output structure to the directory. */
    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    // "composite": true,                     /* Enable project compilation */
    // "tsBuildInfoFile": "./",               /* Specify file to store incremental compilation information */
    // "removeComments": true,                /* Do not emit comments to output. */
    // "noEmit": true,                        /* Do not emit outputs. */
    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */

    /* Strict Type-Checking Options */
    "strict": true,                           /* Enable all strict type-checking options. */
    // "noImplicitAny": true,                 /* Raise error on expressions and declarations with an implied 'any' type. */
    // "strictNullChecks": true,              /* Enable strict null checks. */
    // "strictFunctionTypes": true,           /* Enable strict checking of function types. */
    // "strictBindCallApply": true,           /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
    // "strictPropertyInitialization": true,  /* Enable strict checking of property initialization in classes. */
    // "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */

    /* Additional Checks */
    // "noUnusedLocals": true,                /* Report errors on unused locals. */
    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */

    /* Module Resolution Options */
    "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
    // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
    // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
    // "typeRoots": [],                       /* List of folders to include type definitions from. */
    // "types": [],                           /* Type declaration files to be included in compilation. */
    // "allowSyntheticDefaultImports": true,  /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
    "esModuleInterop": true,                  /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
    // "preserveSymlinks": true,              /* Do not resolve the real path of symlinks. */
    // "allowUmdGlobalAccess": true,          /* Allow accessing UMD globals from modules. */

    /* Source Map Options */
    // "sourceRoot": "",                      /* Specify the location where debugger should locate TypeScript files instead of source locations. */
    // "mapRoot": "",                         /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */

    /* Experimental Options */
    "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */

    /* Advanced Options */
    "skipLibCheck": true,                     /* Skip type checking of declaration files. */
    "forceConsistentCasingInFileNames": true  /* Disallow inconsistently-cased references to the same file. */
  }
}

```

2. Create vue file

```vue
<template>
  <button :class="className" v-on:click="onClick">
    <slot></slot>
  </button>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";

const ButtonBase = Vue.extend({
  props: {
    block: {
      type: Boolean,
      default() {
        return false;
      },
    },
    variant: {
      type: String,
      default() {
        return "primary";
      },
    },
  },
})

@Component
class Button extends ButtonBase {
  get className(): string {
    let classes: string[] = ["btn"];

    if (this.$props.block) {
      classes.push("btn-block");
    }

    classes.push(`btn-${this.$props.variant}`);

    return classes.join(" ");
  }

  onClick(event: MouseEvent) {
    this.$emit("click", event);
  }
}

export default Button;
</script>
```

# Edit

When I change the file to the following, it does fix the issue

```vue
<template>
  <button :class="className" v-on:click="onClick">
    <slot></slot>
  </button>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";

@Component({
 props: {
    block: {
      type: Boolean,
      default() {
        return false;
      },
    },
    variant: {
      type: String,
      default() {
        return "primary";
      },
    },
  },
})
class Button extends Vue {
  get className(): string {
    let classes: string[] = ["btn"];

    if (this.$props.block) {
      classes.push("btn-block");
    }

    classes.push(`btn-${this.$props.variant}`);

    return classes.join(" ");
  }

  onClick(event: MouseEvent) {
    this.$emit("click", event);
  }
}

export default Button;
</script>
```

**Comment by yoyo930021** [comment]
I didn't know they could mix. 😱
The crash is easy problem,
but get `Vue.extend()` part info is hard.

When class extends `Vue.extend()` of this case, I only can get
```
// block
{
   type: Boolean,
   default() {
      return false;
   },
}
// variant
variant: {
  type: String,
  default() {
    return "primary";
  },
}
```
if have any other like `methods` `computed`, I can't detect it. =_=
I will find some solutions for this problem.

**Comment by Zehua-Chen** [comment]
I learned it from here [Props Definition](https://class-component.vuejs.org/guide/props-definition.html)

**Comment by yoyo930021** [comment]
> I learned it from here [Props Definition](https://class-component.vuejs.org/guide/props-definition.html)

I have flipped to this document, thanks.

**Comment by yoyo930021** [comment]
I fixed crash and open a new issue about support this case.
#2261

**Comment by ascott18** [comment]
I'm running into the same error, also with Typescript 4.0.2:

```
[Error - 3:53:06 PM] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'kind' of undefined
  Code: -32603 
```

This happens even in a dead-simple class-based SFC:
``` vue
<template> 

</template>

<script lang="ts">
import { Vue, Component, Prop, Watch } from "vue-property-decorator";

@Component({
  components: {}
})
export default class extends Vue {}
</script>

<style lang="scss"></style>

```

**Comment by octref** [comment]
Bugfixes will be released in 0.27.3 soon.
